### PR TITLE
sjawhar-legion-273: add write-existing-file guard hook

### LINKED ISSUE #273: OMO Replacement: Write-Existing-File Guard Hook (T4)
## Context
Part of #200 — Replace oh-my-opencode (OMO) with opencode-legion.

**Wave 1 — Foundation** | Blocks: T22 | Blocked By: None

## Problem Statement
Prevent agents from overwriting existing files without reading them first — a common cause of data loss.

## What to Do
- `tool.execute.before` hook for Write tool calls
- Track which files have been Read in the current session
- If Write targets an existing file not previously Read in this session, add a warning/reminder
- Do NOT hard-block — just warn
- Clean up file tracking on session.deleted

## Must NOT Do
- Do NOT hard-block writes (soft warning only)
- Do NOT track reads across sessions

## Existing Infrastructure
- `src/hooks/` — 12 hooks, consistent patterns
- `src/hooks/utils.ts` — shared utilities
- `src/hooks/subagent-question-blocker.ts` — good pattern for per-session tracking via Set

## Acceptance Criteria
- [ ] TDD: failing test first, then implementation
- [ ] Write to existing file that was NOT read → adds warning to tool result
- [ ] Write to existing file that WAS read → passes cleanly (no warning)
- [ ] Write to new file (doesn't exist yet) → passes cleanly
- [ ] File tracking cleaned up on session.deleted

## Testing Infrastructure
- `bun test` available
- Follow patterns from existing hook tests

## QA
- Write-without-read → warning present
- Write-after-read → no warning
- Write new file → no warning
Evidence: `.sisyphus/evidence/task-4-write-guard.txt`

## Commit Message
`feat(legion): add write-existing-file guard hook`

## Skills
None required.

## Routing Hints
- Complexity: quick
- Estimated implementers: 1

### PR DESCRIPTION
**Pull request by sjawhar**
Closes #273

## Summary

Adds a write-existing-file guard hook that warns agents when they attempt to write/edit an existing file without reading it first — preventing a common cause of data loss.

### Implementation

- **`createWriteExistingFileGuard`** — factory hook (Pattern B) with `tool.execute.before`, `tool.execute.after`, and `event` lifecycle handlers
- **`createReadFileRegistry`** — per-session `Map<string, Set<string>>` tracking which files each session has Read
- **`tool.execute.after`** — intercepts `read`/`mcp_read` tool calls to register file paths
- **`tool.execute.before`** — intercepts `write`/`edit`/`mcp_write`/`mcp_edit` tool calls, checks if file exists on disk AND hasn't been read, adds soft warning to `output.args.__warning`
- **`event`** — cleans up per-session tracking on `session.deleted`

### Design decisions
- **Soft warning only** (no hard-block) — sets `output.args.__warning`, never throws
- **Per-session isolation** — reads tracked independently per session
- **Path normalization** — `node:path.normalize()` prevents mismatches from `./` or `../` in paths
- **Only `read`/`mcp_read` count as reads** — `grep`, `glob`, `bash cat` don't count (issue spec: "Track which files have been Read")
- Both `createWriteExistingFileGuard` and `createReadFileRegistry` are re-exported from plugin index, satisfying the OMO replacement matrix contract tests (Area 4)

### Tests (34 total)
- Core behavior: write-without-read warns, write-after-read passes, new-file passes
- Per-session isolation: sessions don't share read state
- Session cleanup: `session.deleted` clears tracking (both `sessionID` and `info.id` formats)
- Edge cases: missing filePath, empty sessionID, path normalization, tool name aliases
- Smoke tests: realistic agent workflow scenarios (implement cycle, parallel sessions, session lifecycle, large refactoring, MCP tool names)
- Registry export tests: standalone registry API

### Verification
- `bunx biome check packages/opencode-plugin/` — clean
- `bunx tsc --noEmit -p packages/opencode-plugin/tsconfig.json` — no errors
- `bun test packages/opencode-plugin/` — 415 pass, 21 fail (all pre-existing OMO matrix)
- OMO matrix Area 4: `createWriteExistingFileGuard` and `createReadFileRegistry` tests now pass

### CODE REVIEW

**Comment by sjawhar** [comment]
## Review Summary

**CRITICAL (P1):** 1 issue
**IMPORTANT (P2):** 1 issue
**MINOR (P3):** 2 suggestions

**Verdict: Needs changes** — merge conflict with main, and one vacuous test.

---

### Acceptance Criteria Checklist

| Criterion | Status | Evidence |
|-----------|--------|----------|
| TDD: failing test first, then implementation | PASS | Test file covers all acceptance criteria. 34 tests, 52 assertions. |
| Write to existing file NOT read → warning | PASS | Lines 38-79 of test, lines 100-131 of impl |
| Write to existing file WAS read → no warning | PASS | Lines 83-155 of test |
| Write to new file → no warning | PASS | Lines 158-175 of test, line 117 of impl (existsSync check) |
| Session cleanup on session.deleted | PASS | Lines 227-347 of test, both sessionID and info.id formats |
| No hard-block (soft warning only) | PASS | Never throws, sets output.args.__warning |
| No cross-session tracking | PASS | Per-session Map<string, Set<string>>, isolation tested |

### Code Quality Assessment

**Strengths:**
- Follows the factory function pattern exactly per `docs/solutions/architecture-patterns/opencode-plugin-hook-authoring-patterns.md`
- Correctly reads filePath from `output.args` in `tool.execute.before` and from `input.args` in `tool.execute.after` (matches hook contract)
- Reuses shared utilities (`isRecord`, `resolveSessionID`) from `hooks/utils.ts`
- Path normalization via `node:path.normalize()` prevents path format mismatches
- Clean index.ts wiring: all 3 hook points registered (event, tool.execute.before, tool.execute.after), export added
- Only `read`/`mcp_read` count as reads (design decision per issue spec, correctly excludes grep/glob/bash)
- Tool names lowercased before comparison — handles case variations

**Issues found — see line-level comments:**

1. **P1: PR has merge conflicts with main.** `gh pr view` reports `CONFLICTING` mergeability. Implementer must rebase before merge.

2. **P2: Vacuous test at line 102** — "passes cleanly when file was globbed before writing" has NO assertions. The test body sets up state then ends with internal comments debating the design. The test title is also misleading: it implies glob-then-write passes cleanly, but the actual implementation does NOT count glob as a read (correctly). This test passes vacuously and should either be removed or converted to assert the correct behavior (glob does NOT satisfy the read requirement → write should warn).

3. **P3: `implement.json` missing trailing newline** — minor, JSON file lacks trailing newline (`\ No newline at end of file` in diff).

4. **P3: `tool.execute.before` has a subtle double-check pattern** — filePath is extracted from `output.args`, then `existsSync` is called on the normalized path, but `registry.hasRead` is also called with the same normalized path. The normalization is applied correctly in both places, so this is fine. Just noting the flow is clear.

### Evidence
- CI status: lint pass (8s), test pass (56s), typecheck pass (16s)
- Mergeability: CONFLICTING
- Test run: 34 pass, 0 fail, 52 expect() calls
- Bot review comments: None

File: packages/opencode-plugin/src/hooks/__tests__/write-existing-file-guard.test.ts [hunk 1]
```diff
@@ -0,0 +1,740 @@
...
+        makeAfterInput("glob", "s-1", { pattern: "**/*.ts" }),
+        makeAfterOutput("/src/foo.ts\n/src/bar.ts")
+      );
+
+      // Write to one of the globbed files — should pass without warning
+      // Note: glob doesn't count as "reading" the file content. Only read/grep do.
+      // Actually, re-reading the issue: "Track which files have been Read"
+      // Glob lists files, doesn't read content. Only Read tool reads.
+      // Let me test Read specifically.
+    });
```
Comment (sjawhar) [inline packages/opencode-plugin/src/hooks/__tests__/write-existing-file-guard.test.ts:119]: **[P2 — IMPORTANT]:** This test has NO assertions. It sets up state (glob after-hook), then ends with internal comments debating whether glob should count as a read. The test passes vacuously (empty test body = pass). The title "passes cleanly when file was globbed" is misleading — the actual implementation correctly does NOT count glob as a read.

Fix: Either remove this test, or convert it to assert the correct behavior:
```typescript
it("glob does NOT count as reading — writing after glob still warns", () => {
  const guard = createWriteExistingFileGuard();
  guard["tool.execute.after"](
    makeAfterInput("glob", "s-1", { pattern: "**/*.ts" }),
    makeAfterOutput("/src/foo.ts")
  );
  const output = makeBeforeOutput({ filePath: "/src/foo.ts", content: "x" });
  guard["tool.execute.before"](
    makeBeforeInput("write", "s-1", { filePath: "/src/foo.ts" }),
    output
  );
  expect(output.args.__warning).toBeDefined(); // glob ≠ read
});
```

File: .legion/implement.json [hunk 2]
```diff
@@ -1,18 +1,17 @@
...
+    "Only read/mcp_read count as reads per issue spec, not grep/glob/bash — important design decision to prevent false sense of safety"
   ],
   "deviations": [],
   "openQuestions": [],
   "subPlanningNeeded": false,
   "schemaVersion": 1,
   "phase": "implement",
-  "completed": "2026-04-13T15:33:38.242Z"
+  "completed": "2026-04-13T16:58:37.811Z"
 }
```
Comment (sjawhar) [inline .legion/implement.json:17]: **[P3 — MINOR]:** Missing trailing newline. JSON files should end with a newline per POSIX convention.

File: packages/opencode-plugin/src/hooks/__tests__/write-existing-file-guard.test.ts (hunk 1 above)
Comment (sjawhar) [reply packages/opencode-plugin/src/hooks/__tests__/write-existing-file-guard.test.ts:119]: Fixed. Replaced the vacuous test with a proper assertion: glob triggers `tool.execute.after` but does NOT register as a read, so a subsequent write to the same file correctly produces a warning. Test now has `expect(output.args.__warning).toBeDefined()` confirming glob != read.

File: .legion/implement.json (hunk 2 above)
Comment (sjawhar) [reply .legion/implement.json:17]: Fixed. The conflict resolution rewrote the file with a trailing newline.

**Comment by sjawhar** [comment]
All review feedback addressed:

**P1 (merge conflicts):** Rebased onto main, resolved conflicts in `.legion/implement.json` and `.legion/review.json`. Clean chain on top of main.

**P2 (vacuous test):** Replaced the empty glob test with a proper assertion confirming glob does NOT count as a read — writing after glob still produces a warning. Test now has `expect(output.args.__warning).toBeDefined()`. Test count stays at 34 but expect() calls increased from 52 to 53.

**P3 (trailing newline):** Fixed via conflict resolution rewrite.

CI green: lint (8s), typecheck (15s), test (57s).

**Comment by sjawhar** [comment]
## Re-Review Summary

**CRITICAL (P1):** 0 issues
**IMPORTANT (P2):** 0 issues
**MINOR (P3):** 0 suggestions

**Verdict: Approved** — all prior review feedback addressed, no new issues found.

---

### Prior Feedback Resolution

| Prior Finding | Status | Evidence |
|--------------|--------|---------|
| P1: Merge conflicts with main | RESOLVED | Mergeability now `MERGEABLE`. Conflicts in `.legion/` handoff files resolved via rebase. |
| P2: Vacuous test (glob-before-write) | RESOLVED | Test at line 102 now asserts `expect(output.args.__warning).toBeDefined()` — correctly confirms glob does NOT count as a read. Test title updated to match behavior. expect() count increased from 52 to 53. |
| P3: Missing trailing newline in implement.json | RESOLVED | File now ends with trailing newline (confirmed via conflict resolution rewrite). |

### Acceptance Criteria Checklist

| Criterion | Status | Evidence |
|-----------|--------|---------|
| TDD: failing test first, then implementation | PASS | 34 tests, 53 expect() calls, all passing |
| Write to existing file NOT read → warning | PASS | Tests lines 38-79; impl lines 100-131 |
| Write to existing file WAS read → no warning | PASS | Tests lines 83-159; impl lines 119-120 |
| Write to new file → no warning | PASS | Test lines 162-178; impl line 117 (existsSync) |
| Session cleanup on session.deleted | PASS | Tests lines 230-350; both sessionID and info.id formats |
| No hard-block (soft warning only) | PASS | Test lines 70-79; impl never throws |
| No cross-session tracking | PASS | Tests lines 180-228; per-session Map isolation |

### Evidence
- CI status: lint pass (6s), test pass (59s), typecheck pass (18s)
- Mergeability: MERGEABLE
- Local test run: 34 pass, 0 fail, 53 expect() calls
- Bot review comments: None

**Comment by sjawhar** [comment]
Closing — .legion/ conflict cascade unresolvable after multiple attempts. Will re-implement from clean main. Code is written and smoke-tested.
